### PR TITLE
Subscriptions. 28. Intercept Privacy Pro URL + Navigation Update Ship review fixes

### PR DESCRIPTION
**Pull request by afterxleep**
Task/Issue URL:  https://app.asana.com/0/0/1206983541370659/f
Tech Design URL:
CC:

**Description**:
- Implements a URLInterceptor class to define custom actions when visiting specific URLs
- Other UI fixes based on Ship Review for navigation Updates
- Minor updates to cancellable removals and deInit cleanUp

**Steps to test this PR**:
1.  Go to any website
2. Go to https://sbarag.duckduckgo.com/pro 
3. Go back until you get to the browser, you should be on the first page you loaded
4. Open a new tab and go to https://codepen.io/marcosholgado/pen/PogEJrM
5. Click on Click Here
6. You should be taken to the native flow
7. Go back to the browser, you should be in the same page as before
8. Long click on click here and open in a new tab
9. You should be taken to the native flow
10. Go back to the browser, you should be in the same page
11. Long click on click here and open in a new background tab
12. You should be taken to the native flow

### CODE REVIEW

File: DuckDuckGo/SettingsSubscriptionView.swift [hunk 1]
```diff
@@ -100,6 +100,8 @@ struct SettingsSubscriptionView: View {
                            isActive: $isShowingRestoreFlow,
                            label: { SettingsCellView(label: UserText.settingsPProIHaveASubscription ) })
         }
+
+
```
Comment (miasma13) [inline DuckDuckGo/SettingsSubscriptionView.swift]: Whitespace

File: DuckDuckGo/Subscription/AsyncHeadlessWebview/AsyncHeadlessWebViewModel.swift [hunk 2]
```diff
@@ -34,7 +34,11 @@ final class AsyncHeadlessWebViewViewModel: ObservableObject {
     
     @Published var scrollPosition: CGPoint = .zero
     @Published var url: URL?
-    @Published var canGoBack: Bool = false
+    @Published var canGoBack: Bool = false {
+        didSet {
+            print("CanGoback \(canGoBack)")
```
Comment (miasma13) [inline DuckDuckGo/Subscription/AsyncHeadlessWebview/AsyncHeadlessWebViewModel.swift]: Leftover print

File: DuckDuckGo/Subscription/UserScripts/SubscriptionPagesUseSubscriptionFeature.swift [hunk 3]
```diff
@@ -417,6 +417,7 @@ final class SubscriptionPagesUseSubscriptionFeature: Subfeature, ObservableObjec
         onSetSubscription = nil
         onActivateSubscription = nil
         onBackToSettings = nil
+
```
Comment (miasma13) [inline DuckDuckGo/Subscription/UserScripts/SubscriptionPagesUseSubscriptionFeature.swift]: Whitespace

**Comment by afterxleep** [comment]
@federicocappelli  Most of this PR is about cleaning up methods, deInit objects and various labels and conditional titles/strings.

The relevant changes are the following:
- [New URL Interceptor class](https://github.com/duckduckgo/iOS/pull/2670/files#diff-eb4c7d1c470401446ec7731d527db061518d5f0f1cb6abc28f0fb6577d9e5b36)
- [Changes to TabViewController WKWebviewDelegate to implement URL interceptor](https://github.com/duckduckgo/iOS/pull/2670/files#diff-23f45a546b44732f442ecaff49aa3d7bbdb120eadd9cf312fc61b8d5181b6ad0)

File: DuckDuckGo/TabURLInterceptor.swift [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        
+        guard let matchingURL = urlToIntercept(tld: urlTLD, path: components.path) else {
+            return true
+        }
+        
+        return Self.handleURLInterception(url: matchingURL.id)
+        
+    }
+}
+
```
Comment (federicocappelli) [inline DuckDuckGo/TabURLInterceptor.swift:63]: too many \n?

File: DuckDuckGo/TabURLInterceptor.swift [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import BrowserServicesKit
+import Common
+import Subscription
```
Comment (federicocappelli) [inline DuckDuckGo/TabURLInterceptor.swift:23]: General note on the class:
From my understanding, the rush for the release is over and I would expect every new feature to be Unit Tested.
I realise that this class is very simple, so It should require very simple tests :)

File: DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift [hunk 6]
```diff
@@ -298,3 +317,16 @@ final class SubscriptionFlowViewModel: ObservableObject {
     
 }
 #endif
+
+// TODO: Move to BSK later
+extension URL {
```
Comment (federicocappelli) [inline DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift]: fileprivate? we don't want this to leak to the entire app right?

**Review by federicocappelli — COMMENTED** [review]
Sew comments and requests

File: DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift [hunk 7]
```diff
@@ -319,7 +319,7 @@ final class SubscriptionFlowViewModel: ObservableObject {
 #endif
 
 // TODO: Move to BSK later
-extension URL {
+private extension URL {
```
Comment (federicocappelli) [inline DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift:322]: this still exposes the extension to the entire app, fileprivate?

File: DuckDuckGoTests/TabURLInterceptorTests.swift [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    }
+    
+    func testNotificationForInterceptedPrivacyProPath() {
+        let expectation = self.expectation(forNotification: .urlInterceptPrivacyPro, object: nil, handler: nil)
+        
+        let url = URL(string: "https://duckduckgo.com/pro")!
+        let canNavigate = urlInterceptor.allowsNavigatingTo(url: url)
+        
+        // Simulate purchase allowance
+        SubscriptionPurchaseEnvironment.canPurchase = true
```
Comment (federicocappelli) [inline DuckDuckGoTests/TabURLInterceptorTests.swift]: Not for this PR but this kind of global state will not work in future

File: DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift (hunk 7 above)
Comment (afterxleep) [reply DuckDuckGo/Subscription/ViewModel/SubscriptionFlowViewModel.swift:322]: whoops, typo..

File: DuckDuckGoTests/TabURLInterceptorTests.swift (hunk 8 above)
Comment (afterxleep) [reply DuckDuckGoTests/TabURLInterceptorTests.swift]: yea.. sadly there's no clear way to mock or substitute for testing on this one.  We could wrap in protocol for this test, but this should probably be rewritten as part of the Unit testing efforts.

**Review by federicocappelli — APPROVED** [review]
LGTM

**Pull request merged**